network/src/validator_network/network_builder.rs

```diff
@@ -517,22 +529,31 @@ impl NetworkBuilder {
             let (discovery_network_tx, discovery_network_rx) = network_provider.add_discovery(
                 vec![ProtocolId::from_static(DISCOVERY_DIRECT_SEND_PROTOCOL)],
             );
-            let discovery = Discovery::new(
-                self.peer_id,
-                vec![self
-                    .advertised_address
-                    .clone()
-                    .unwrap_or_else(|| self.addr.clone())],
-                signer,
-                self.seed_peers.clone(),
-                self.trusted_peers.clone(),
-                Interval::new_interval(Duration::from_millis(self.discovery_interval_ms)).fuse(),
-                discovery_network_tx,
-                discovery_network_rx,
-                net_conn_mgr_reqs_tx.take().unwrap(),
-                Duration::from_millis(self.discovery_msg_timeout_ms),
-            );
-            self.executor.spawn(discovery.start());
+            let peer_id = self.peer_id;
+            let addrs = vec![self
+                .advertised_address
+                .clone()
+                .unwrap_or_else(|| self.addr.clone())];
+            let seed_peers = self.seed_peers.clone();
+            let trusted_peers = self.trusted_peers.clone();
+            let discovery_interval_ms = self.discovery_interval_ms;
+            let discovery_msg_timeout_ms = self.discovery_msg_timeout_ms;
+            let f = async move {
+                let discovery = Discovery::new(
+                    peer_id,
+                    addrs,
+                    signer,
+                    seed_peers,
+                    trusted_peers,
+                    interval(Duration::from_millis(discovery_interval_ms)).fuse(),
+                    discovery_network_tx,
+                    discovery_network_rx,
+                    net_conn_mgr_reqs_tx.take().unwrap(),
+                    Duration::from_millis(discovery_msg_timeout_ms),
+                );
+                discovery.start().await
+            };
+            self.executor.spawn(f);
```

Comment:
man, this really sucks...

I wonder if one way we could get around the new limitation where we can't construct timers/intervals outside of a tokio context would be to create a custom, lazy wrapper. This wrapper might look like

```rust
struct MyInterval {
    interval_duration: Duration,
    inner: Option<tokio::timer::Interval>,
}

impl MyInterval {
    pub fn new(deadline_interval: Duration) -> Self {
        Self {
            deadline_interval,
            inner: None,
        }
    }
}

impl Stream for MyInterval {
    type Item = ();

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
        if self.inner.is_none() {
            self.inner = interval(self.interval_duration).fuse();
        }

        self.inner.unwrap().poll_next(&mut cx)
    }
}
```

Reply:
Yeah this was a very unfortunate find...